README.md
# Seconds

[![Tests](https://github.com/adevade/seconds/workflows/Tests/badge.svg)](https://github.com/adevade/seconds)
[![Latest Version on Packagist](https://img.shields.io/packagist/v/adevade/seconds.svg)](https://packagist.org/packages/adevade/seconds)
[![Total Downloads](https://img.shields.io/packagist/dt/adevade/seconds.svg)](https://packagist.org/packages/adevade/seconds)

Helpers for converting time to seconds.

## Installation

You can install the package via composer:

```bash
composer require adevade/seconds
```

### Supported PHP versions

| Package | PHP        |
| ------- | ---------- |
| ^1.0    | ^7.2       |
| ^2.1    | ^7.2\|^8.0 |
| ^3.0    | ^7.4\|^8.0 |
| ^4.0    | ^8.0       |

## Usage

```php
use Adevade\Seconds;

Seconds::fromMinutes(2); // returns => (int) 120
```

### Available methods

```php
Seconds::fromMinutes($minutes = 5);
Seconds::fromHours($hours = 12);
Seconds::fromDays($days = 4);
Seconds::fromWeeks($weeks = 2);
Seconds::fromMonths($months = 6);
Seconds::fromYears($years = 2);

Seconds::fromMinute();
Seconds::fromHour();
Seconds::fromDay();
Seconds::fromWeek();
Seconds::fromMonth();
Seconds::fromYear();
```

### Available constants

```php
Seconds::MINUTE;
Seconds::HOUR;
Seconds::DAY;
Seconds::WEEK;
Seconds::MONTH;
Seconds::YEAR;
```

> Months have an average length of 30.42 days.\
> Years have an average length of 365.24 days.

## Credits

- [Andréas Lundgren](https://github.com/adevade)

Idea came from a tweet by [@LasseRafn](https://twitter.com/LasseRafn). Thanks!


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [Unreleased]

## [4.0.1] - 2023-05-25

### Changed

- Refactor internals.

## [4.0.0] - 2022-10-03

### Removed

- Drop support for PHP 7.

## [3.0.0] - 2021-10-25

### Removed

- Drop support for PHP <= 7.3.
- Drop support for PHPUnit 8.

## [2.1.0] - 2021-04-16

### Added

- Support for PHP 8.0.

### Changed

- Exception messages use correct variable names.

## [2.0.0] - 2020-04-17

:boom: Breaking changes! New namespace. Run `composer dump-autoload` after update.

### Changed

- Change namespace to top level `Adevade`. Use `Adevade\Seconds::class` instead of `Adevade\Seconds\Seconds::class` in ^2.0.

## [1.1.0] - 2020-04-17

### Added

- Added support for PHPUnit `^8|^9`.

## [1.0.2] - 2020-02-28

### Changed

- Better variable names for DocBlock helpers.

## [1.0.1] - 2020-02-08

### Added

- DocBlock helpers for IDEs.

### Fixed

- Throw exceptions if method does not exist or if parameter is invalid.

## [1.0.0] - 2020-02-06

- Initial release

[unreleased]: https://github.com/adevade/seconds/compare/4.0.1...HEAD
[4.0.1]: https://github.com/adevade/seconds/compare/4.0.0...4.0.1
[4.0.0]: https://github.com/adevade/seconds/compare/3.0.0...4.0.0
[3.0.0]: https://github.com/adevade/seconds/compare/2.1.0...3.0.0
[2.1.0]: https://github.com/adevade/seconds/compare/2.0.0...2.1.0
[2.0.0]: https://github.com/adevade/seconds/compare/1.1.0...2.0.0
[1.1.0]: https://github.com/adevade/seconds/compare/1.0.2...1.1.0
[1.0.2]: https://github.com/adevade/seconds/compare/1.0.1...1.0.2
[1.0.1]: https://github.com/adevade/seconds/compare/1.0.0...1.0.1
[1.0.0]: https://github.com/adevade/seconds/releases/tag/1.0.0


tests/ExampleTest.php
<?php

namespace 035media\Seconds\Tests;

use PHPUnit\Framework\TestCase;

class ExampleTest extends TestCase
{
    /** @test */
    public function true_is_true()
    {
        $this->assertTrue(true);
    }
}


tests/SecondsTest.php
<?php

namespace Adevade\Seconds\Tests;

use Adevade\Seconds;
use BadMethodCallException;
use InvalidArgumentException;
use PHPUnit\Framework\TestCase;

class SecondsTest extends TestCase
{
    /** @test */
    function it_converts_from_minutes_to_seconds()
    {
        $this->assertSame(60, Seconds::fromMinutes(1));
        $this->assertSame(120, Seconds::fromMinutes(2));
        $this->assertSame(0, Seconds::fromMinutes(0));

        $this->assertSame(60, Seconds::fromMinute());
        $this->assertSame(60, Seconds::fromMinute(999));
    }

    /** @test */
    function it_converts_from_hours_to_seconds()
    {
        $this->assertSame(3600, Seconds::fromHours(1));
        $this->assertSame(7200, Seconds::fromHours(2));
        $this->assertSame(0, Seconds::fromHours(0));

        $this->assertSame(3600, Seconds::fromHour());
        $this->assertSame(3600, Seconds::fromHour(999));
    }

    /** @test */
    function it_converts_from_days_to_seconds()
    {
        $this->assertSame(86400, Seconds::fromDays(1));
        $this->assertSame(172800, Seconds::fromDays(2));
        $this->assertSame(0, Seconds::fromDays(0));

        $this->assertSame(86400, Seconds::fromDay());
        $this->assertSame(86400, Seconds::fromDay(999));
    }

    /** @test */
    function it_converts_from_weeks_to_seconds()
    {
        $this->assertSame(604800, Seconds::fromWeeks(1));
        $this->assertSame(1209600, Seconds::fromWeeks(2));
        $this->assertSame(0, Seconds::fromWeeks(0));

        $this->assertSame(604800, Seconds::fromWeek());
        $this->assertSame(604800, Seconds::fromWeek(999));
    }

    /** @test */
    function it_converts_from_months_to_seconds()
    {
        $this->assertSame(2628288, Seconds::fromMonths(1));
        $this->assertSame(5256576, Seconds::fromMonths(2));
        $this->assertSame(0, Seconds::fromMonths(0));

        $this->assertSame(2628288, Seconds::fromMonth());
        $this->assertSame(2628288, Seconds::fromMonth(999));
    }

    /** @test */
    function it_converts_from_years_to_seconds()
    {
        $this->assertSame(31556736, Seconds::fromYears(1));
        $this->assertSame(63113472, Seconds::fromYears(2));
        $this->assertSame(0, Seconds::fromYears(0));

        $this->assertSame(31556736, Seconds::fromYear());
        $this->assertSame(31556736, Seconds::fromYear(999));
    }

    /** @test */
    function it_throws_an_exception_if_no_parameter_is_given()
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage(
            'Adevade\Seconds::fromMinutes(int $minutes) expects an integer.'
        );

        Seconds::fromMinutes();
    }

    /** @test */
    function it_throws_an_exception_if_parameter_is_not_an_integer()
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage(
            'Adevade\Seconds::fromDays(int $days) expects an integer.'
        );

        Seconds::fromDays('4');
    }

    /** @test */
    function it_throws_an_exception_if_method_does_not_exist()
    {
        $this->expectException(BadMethodCallException::class);
        $this->expectExceptionMessage(
            'Adevade\Seconds::invalidMethod() does not exist.'
        );

        Seconds::invalidMethod(12);
    }
}


src/Seconds.php
<?php

namespace Adevade;

use BadMethodCallException;
use InvalidArgumentException;

/**
 * @method static int fromMinute()
 * @method static int fromMinutes(int $minutes)
 * @method static int fromHour()
 * @method static int fromHours(int $hours)
 * @method static int fromDay()
 * @method static int fromDays(int $days)
 * @method static int fromWeek()
 * @method static int fromWeeks(int $weeks)
 * @method static int fromMonth()
 * @method static int fromMonths(int $months)
 * @method static int fromYear()
 * @method static int fromYears(int $years)
 */
class Seconds
{
    public const MINUTE = 60;

    public const HOUR = 3600;

    public const DAY = 86400;

    public const WEEK = 604800;

    public const MONTH = 2628288; // 30.42 days

    public const YEAR = 31556736; // 365.24 days

    /**
     * Return the calculated seconds.
     *
     * @throws \InvalidArgumentException
     */
    public static function __callStatic(string $method, array $parameters): int
    {
        if (static::isSingular($method)) {
            return static::getConstantFromMethodName($method);
        }

        if (! static::hasValidParameters($parameters)) {
            throw new InvalidArgumentException(
                sprintf(
                    '%s::%s(int $%s) expects an integer.',
                    static::class,
                    $method,
                    str_replace('from', '', strtolower($method))
                )
            );
        }

        return static::getConstantFromMethodName($method) * $parameters[0];
    }

    /**
     * Check if method is singular or plural.
     */
    protected static function isSingular(string $method): bool
    {
        return ! str_ends_with($method, 's');
    }

    /**
     * Check if the passed parameters are present and valid.
     */
    protected static function hasValidParameters(array $parameters): bool
    {
        return isset($parameters[0]) && is_int($parameters[0]);
    }

    /**
     * Get constant from method name.
     *
     * @throws \BadMethodCallException
     */
    protected static function getConstantFromMethodName(string $method): int
    {
        $constant = preg_replace('/^from/', '', $method);
        $constant = preg_replace('/s$/', '', $constant);
        $constant = strtoupper($constant);

        if (! defined("static::{$constant}")) {
            throw new BadMethodCallException(
                sprintf('%s::%s() does not exist.', static::class, $method)
            );
        }

        return constant("static::{$constant}");
    }
}
